### PR TITLE
Add relative monad

### PR DESCRIPTION
**Pull request by shmookey**
Relative monads as described by "Monads need not be endofunctors" http://arxiv.org/pdf/1412.7148.pdf

### CODE REVIEW

File: project/plugins.sbt [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
 resolvers += Resolver.url("commbank-releases-ivy", new URL("http://commbank.artifactoryonline.com/commbank/ext-releases-local-ivy"))(Patterns("[organization]/[module]_[scalaVersion]_[sbtVersion]/[revision]/[artifact](-[classifier])-[revision].[ext]"))
 
-val uniformVersion = "1.2.2-20150429235901-67f611b"
+val uniformVersion = "1.2.7-20151007231119-4b82c69"
```
Comment (vineethvarghese) [inline project/plugins.sbt]: Can we bump up to the latest uniform version ?

Comment (shmookey) [inline project/plugins.sbt]: Done

File: core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      NrelR.rPoint(MrelN.rPoint(expr))
+
+    def rBind[A, B](ra: R[A])(f: M[A] => R[B]): R[B] =
+      NrelR.rBind(ra)((na: N[A]) => R.join(NrelR.rPoint(
+        MrelN.rBind(na)(ma => N.point(f(ma)))
+      )))
+  }
+
+  def rMap[A, B, M[_], N[_]](na: N[A])(f: M[A] => M[B])(
+    implicit M: Monad[M], N: Monad[N], MrelN: RelMonad[M, N]
```
Comment (rowandavies) [inline core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala]: I think we don't need the M: Monad[M] and N: Monad[N] implicits anymore.  It's more general without them, but so far I don't think we have any specific uses of that generality.

Also, add a comment like:  `/** Relative map, handy for lifting functions from M to N */` 

(I guess we have a relative functor, but that doesn't seem to be a concept mentioned anywhere in the literature.  It seems like a generalisation of a natural transformation, or perhaps it is the same.)

File: core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package au.com.cba.omnia.omnitool
+
+import scalaz.{Monad, Id}
+import scalaz.Scalaz.Id
+
+/** Base trait for relative monads. Defines `M` relative to `R`.
```
Comment (rowandavies) [inline core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala]: I'd say something like:   `Relative monad instances relate two functors in a way generalising the relationship between pure expressions and ordinary monads.`

File: core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  *
+  * Typically a number of `R`s will be made relative to a single `M` such that
+  * operations of each type `R[A]` can be viewed and composed as an `M[A]`.
+  *
```
Comment (rowandavies) [inline core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala]: I'd say: `Often we compose operations from one or more simple R's by choosing an actual complicated M that has relative instances from each R.`

**Comment by rowandavies** [comment]
:+1: from me.  I'd like to see relative monad laws here at some point though.

File: core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala (hunk 2 above)
Comment (shmookey) [inline core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala]: Yep - compiles fine without those implicits. :)

File: core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala (hunk 3 above)
Comment (shmookey) [inline core/src/main/scala/au/com/cba/omnia/omnitool/RelMonad.scala]: Ah, that's a good way of putting it.

**Comment by rowandavies** [comment]
:+1:  Just one small fix.

**Pull request merged**